File: sorting/java/sorting/src/xyz/zrui/algorithm/sorting/Quick2Sort.java
package xyz.zrui.algorithm.sorting;


import java.util.Random;

/**
 * 双路快速排序
 */
public class Quick2Sort implements Sorting {


    /**
     * 交换数组间的位置
     *
     * @param arr 目标数组
     * @param i   交换位置索引
     * @param j   交换位置索引
     */
    private void swap(Object[] arr, int i, int j) {
        Object t = arr[i];
        arr[i] = arr[j];
        arr[j] = t;
    }


    private int partition2(Comparable[] arr, int l, int r) {

        Random random = new Random();
        int k = random.nextInt(r - l + 1) + l;
        swap(arr, l, k);

        Comparable v = arr[l];

        //arr[l+1....i] <= v    arr[r...j] >=v
        int i = l + 1;
        int j = r;
        while (true) {

            while (i <= r && arr[i].compareTo(v) < 0) {
                i++;
            }

            while (j >= l + 1 && arr[j].compareTo(v) > 0) {
                j--;
            }
            if (i > j) {
                break;
            }
            swap(arr, i, j);
            i++;
            j--;
        }
        // j 为从右往左的最后一个 小于等于v
        swap(arr, j, l);
        return j;

    }

    private void sort(Comparable[] arr, int l, int r) {
        if (l > r) {
            return;
        }

        int p = partition2(arr, l, r);
        sort(arr, l, p - 1);
        sort(arr, p + 1, r);

    }


    @Override
    public void sort(Comparable[] arr) {

        sort(arr, 0, arr.length - 1);

    }

    public static void main(String[] args) {

        Comparable[] arr = SortTestHelper.generateRandomArray(10, 1, 9);
        SortTestHelper.printArray(arr);
        SortTestHelper.testSort(Quick2Sort.class, arr);
    }


}

File: sorting/python/quick_2_sort.py
# -*- coding: utf-8 -*

'''
    两路快速排序
'''
import random
import sort_helper

def __partition2(arr,l,r):
    k = random.randint(l, r)
    arr[l], arr[k] = arr[k], arr[l]

    v = arr[l]


    i, j = l + 1, r

    while True:

        while i <= r and arr[i] < v :
            i += 1
        while j >= l + 1 and arr[j] > v:
            j -= 1

        if i > j:
            break

        arr[i], arr[j] = arr[j], arr[i]
        i += 1
        j -= 1
    arr[l] , arr[j] = arr[j], arr[l]
    return j

def __sort(arr,l,r):

    if l > r:
        return

    p = __partition2(arr,l,r)
    __sort(arr,l, p - 1)
    __sort(arr,p + 1, r)


@sort_helper.test_sort
def sort(arr):
    __sort(arr,0,len(arr) - 1)


if __name__ == '__main__':
    a = sort_helper.generate_randoma_array(1000000,1,99999)
    sort(a)
